File: MiniDB_Projects/Team_OrionDB/src/concurrency/lock_manager.py
import threading
from collections import defaultdict

class LockRequest:
    def __init__(self, txn_id, lock_mode):
        self.txn_id = txn_id
        self.lock_mode = lock_mode  # 'S' or 'X'
        self.event = threading.Event()
        self.aborted = False

class LockManager:
    def __init__(self):
        self.lock = threading.Lock()
        # resource_id -> list of LockRequest (both granted and waiting)
        self.lock_table = defaultdict(list)
        # txn_id -> set of resource_ids locked
        self.txn_locks = defaultdict(set)
        
        # Wait-for graph: txn_id -> set of txn_ids that it is waiting for
        self.wait_for_graph = defaultdict(set)

    def acquire_shared(self, txn_id, resource_id):
        return self._acquire(txn_id, resource_id, 'S')

    def acquire_exclusive(self, txn_id, resource_id):
        return self._acquire(txn_id, resource_id, 'X')

    def _acquire(self, txn_id, resource_id, mode):
        req = LockRequest(txn_id, mode)
        
        with self.lock:
            # Check if this transaction already holds an X lock on the resource
            existing_locks = self.lock_table[resource_id]
            for r in existing_locks:
                if r.txn_id == txn_id and r.event.is_set():
                    if r.lock_mode == 'X' or mode == 'S':
                        # Already holds a sufficient lock
                        return True
                    elif r.lock_mode == 'S' and mode == 'X':
                        # Lock upgrade (simplified: if only this txn holds it, upgrade, otherwise conflict)
                        holders = [hl for hl in existing_locks if hl.event.is_set()]
                        if len(holders) == 1 and holders[0].txn_id == txn_id:
                            holders[0].lock_mode = 'X'
                            return True
            
            # Add request to lock table queue
            existing_locks.append(req)
            
            # Check if request can be granted immediately
            if self._is_compatible(resource_id, req):
                req.event.set()
                self.txn_locks[txn_id].add(resource_id)
                return True
                
            # If not compatible, we must wait. Let's update Wait-For Graph and check for deadlocks.
            holders = [hl for hl in existing_locks if hl.event.is_set() and hl.txn_id != txn_id]
            for h in holders:
                self.wait_for_graph[txn_id].add(h.txn_id)
                
            # Detect deadlock
            if self._has_deadlock():
                # Deadlock detected! We abort the youngest transaction in the cycle.
                # In our design, the transaction with the largest Txn ID is the youngest.
                # Let's clean up our request and return False.
                existing_locks.remove(req)
                self._remove_wait_edges(txn_id)
                return False

        # Wait outside the LockManager lock to avoid blocking other transactions
        # E.g. wait up to 5 seconds to avoid indefinite blocks (lock timeout)
        granted = req.event.wait(timeout=5.0)
        
        with self.lock:
            if not granted:
                # Lock timeout
                if req in self.lock_table[resource_id]:
                    self.lock_table[resource_id].remove(req)
                self._remove_wait_edges(txn_id)
                return False
                
            if req.aborted:
                return False
                
            self.txn_locks[txn_id].add(resource_id)
            return True

    def _is_compatible(self, resource_id, request):
        queue = self.lock_table[resource_id]
        # Any request ahead of us in the queue that has been granted?
        for req in queue:
            if req == request:
                break
            # If there is a granted request that conflicts, not compatible
            if req.event.is_set():
                if request.lock_mode == 'X' or req.lock_mode == 'X':
                    return False
        return True

    def _has_deadlock(self):
        # Run DFS cycle detection on wait_for_graph
        visited = {}  # txn_id -> state: 0 = unvisited, 1 = visiting, 2 = visited
        
        def dfs(node):
            visited[node] = 1  # visiting
            for neighbor in self.wait_for_graph[node]:
                state = visited.get(neighbor, 0)
                if state == 1:
                    return True  # Found cycle!
                elif state == 0:
                    if dfs(neighbor):
                        return True
            visited[node] = 2  # visited
            return False

        for node in list(self.wait_for_graph.keys()):
            if visited.get(node, 0) == 0:
                if dfs(node):
                    return True
        return False

    def _remove_wait_edges(self, txn_id):
        if txn_id in self.wait_for_graph:
            del self.wait_for_graph[txn_id]
        for key in list(self.wait_for_graph.keys()):
            if txn_id in self.wait_for_graph[key]:
                self.wait_for_graph[key].remove(txn_id)

    def release_locks(self, txn_id):
        with self.lock:
            locked_resources = list(self.txn_locks[txn_id])
            for res_id in locked_resources:
                queue = self.lock_table[res_id]
                # Find and remove our granted requests
                for req in list(queue):
                    if req.txn_id == txn_id and req.event.is_set():
                        queue.remove(req)
                
                # Check waiting requests on this resource and grant them
                for req in queue:
                    if not req.event.is_set() and self._is_compatible(res_id, req):
                        req.event.set()
                        
                if not queue:
                    del self.lock_table[res_id]
                    
            del self.txn_locks[txn_id]
            self._remove_wait_edges(txn_id)

File: MiniDB_Projects/Learners/src/tx/lock_manager.cpp
#include "lock_manager.h"
#include <algorithm>
#include <iostream>

std::string LockManager::make_resource_id(const std::string& table, std::pair<int, int> rid) const {
    return table + "|" + std::to_string(rid.first) + "," + std::to_string(rid.second);
}

bool LockManager::acquire_shared(int txn_id, const std::string& table, std::pair<int, int> rid) {
    std::string res_id = make_resource_id(table, rid);
    auto req = std::make_shared<LockRequest>(txn_id, 'S');

    LockGuard lck(mu);
    
    LockEntry& entry = lock_table[res_id];
    
    // Check if already holds S or X
    auto holder_it = entry.holders.find(txn_id);
    if (holder_it != entry.holders.end()) {
        return true;
    }

    // Check conflict: any holder of X
    bool conflict = false;
    for (const auto& holder : entry.holders) {
        if (holder.second == 'X') {
            conflict = true;
            break;
        }
    }

    if (!conflict && entry.waiters.empty()) {
        entry.holders[txn_id] = 'S';
        txn_locks[txn_id].insert(res_id);
        return true;
    }

    // Add to waiters and check deadlock
    entry.waiters.push_back(req);
    if (has_deadlock(txn_id)) {
        entry.waiters.pop_back();
        throw DeadlockException("Deadlock detected! Transaction " + std::to_string(txn_id) + " aborted.");
    }

    // Wait until lock is granted
    auto cv_ptr = req->cv;
    while (!req->granted) {
        cv_ptr->wait(mu);
    }
    return true;
}

bool LockManager::acquire_exclusive(int txn_id, const std::string& table, std::pair<int, int> rid) {
    std::string res_id = make_resource_id(table, rid);
    auto req = std::make_shared<LockRequest>(txn_id, 'X');

    LockGuard lck(mu);
    
    LockEntry& entry = lock_table[res_id];
    
    // Check if already holds X
    auto holder_it = entry.holders.find(txn_id);
    if (holder_it != entry.holders.end()) {
        if (holder_it->second == 'X') {
            return true;
        }
        // Upgrade from S to X if we are the only holder
        if (entry.holders.size() == 1) {
            entry.holders[txn_id] = 'X';
            return true;
        }
    }

    // Conflict if there are any other holders
    bool conflict = !entry.holders.empty();

    if (!conflict && entry.waiters.empty()) {
        entry.holders[txn_id] = 'X';
        txn_locks[txn_id].insert(res_id);
        return true;
    }

    // Add to waiters and check deadlock
    entry.waiters.push_back(req);
    if (has_deadlock(txn_id)) {
        entry.waiters.pop_back();
        throw DeadlockException("Deadlock detected! Transaction " + std::to_string(txn_id) + " aborted.");
    }

    // Wait until lock is granted
    auto cv_ptr = req->cv;
    while (!req->granted) {
        cv_ptr->wait(mu);
    }
    return true;
}

void LockManager::release_locks(int txn_id) {
    LockGuard lck(mu);
    auto it = txn_locks.find(txn_id);
    if (it == txn_locks.end()) {
        return;
    }

    std::unordered_set<std::string> resources = it->second;
    for (const auto& res_id : resources) {
        auto entry_it = lock_table.find(res_id);
        if (entry_it == lock_table.end()) {
            continue;
        }
        LockEntry& entry = entry_it->second;
        entry.holders.erase(txn_id);

        grant_next_waiters(res_id, entry);

        if (entry.holders.empty() && entry.waiters.empty()) {
            lock_table.erase(entry_it);
        }
    }
    txn_locks.erase(it);
}

void LockManager::grant_next_waiters(const std::string& resource_id, LockEntry& entry) {
    if (!entry.holders.empty()) {
        // If current holder is exclusive, nothing can be granted
        for (const auto& holder : entry.holders) {
            if (holder.second == 'X') return;
        }

        // Holders are shared, we can only grant shared locks from the waiters list
        std::vector<std::shared_ptr<LockRequest>> to_grant;
        for (auto req_it = entry.waiters.begin(); req_it != entry.waiters.end(); ) {
            auto req = *req_it;
            if (req->lock_type == 'S') {
                req_it = entry.waiters.erase(req_it);
                entry.holders[req->txn_id] = 'S';
                txn_locks[req->txn_id].insert(resource_id);
                to_grant.push_back(req);
            } else {
                break; // X lock request blocks further grants
            }
        }
        for (auto& req : to_grant) {
            req->granted = true;
            req->cv->notify_all();
        }
        return;
    }

    if (entry.waiters.empty()) {
        return;
    }

    auto next_req = entry.waiters[0];
    if (next_req->lock_type == 'X') {
        entry.waiters.erase(entry.waiters.begin());
        entry.holders[next_req->txn_id] = 'X';
        txn_locks[next_req->txn_id].insert(resource_id);
        next_req->granted = true;
        next_req->cv->notify_all();
    } else {
        std::vector<std::shared_ptr<LockRequest>> to_grant;
        while (!entry.waiters.empty() && entry.waiters[0]->lock_type == 'S') {
            auto req = entry.waiters[0];
            entry.waiters.erase(entry.waiters.begin());
            entry.holders[req->txn_id] = 'S';
            txn_locks[req->txn_id].insert(resource_id);
            to_grant.push_back(req);
        }
        for (auto& req : to_grant) {
            req->granted = true;
            req->cv->notify_all();
        }
    }
}

bool LockManager::has_deadlock(int start_txn_id) {
    std::unordered_map<int, std::unordered_set<int>> graph;
    
    for (const auto& item : lock_table) {
        const LockEntry& entry = item.second;
        std::vector<int> holders;
        for (const auto& holder : entry.holders) {
            holders.push_back(holder.first);
        }
        
        for (const auto& req : entry.waiters) {
            int waiter_id = req->txn_id;
            for (int holder_id : holders) {
                if (waiter_id != holder_id) {
                    graph[waiter_id].insert(holder_id);
                }
            }
            
            for (const auto& prior_req : entry.waiters) {
                if (prior_req == req) break;
                if (req->lock_type == 'X' || prior_req->lock_type == 'X') {
                    graph[waiter_id].insert(prior_req->txn_id);
                }
            }
        }
    }

    // Detect cycle from start_txn_id using DFS
    std::unordered_set<int> visited;
    std::unordered_set<int> rec_stack;

    std::function<bool(int)> dfs = [&](int node) -> bool {
        visited.insert(node);
        rec_stack.insert(node);

        auto it = graph.find(node);
        if (it != graph.end()) {
            for (int neighbor : it->second) {
                if (visited.find(neighbor) == visited.end()) {
                    if (dfs(neighbor)) return true;
                } else if (rec_stack.find(neighbor) != rec_stack.end()) {
                    return true;
                }
            }
        }

        rec_stack.erase(node);
        return false;
    };

    return dfs(start_txn_id);
}
